tests/components/bsblan/test_config_flow.py

```diff
@@ -112,3 +113,319 @@ async def test_user_device_exists_abort(
+    assert result["step_id"] == "discovery_confirm"
+
+    # Now complete the discovery by providing credentials and connecting
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {
+            CONF_USERNAME: "admin",
+            CONF_PASSWORD: "secret",
+        },
+    )
+
+    await hass.async_block_till_done()
+
+    assert result2["type"] == FlowResultType.CREATE_ENTRY
+    assert result2["title"] == "00:80:41:19:69:90"  # MAC from fixture file
+    assert result2["data"] == {
+        CONF_HOST: "10.0.2.60",
+        CONF_PORT: 80,
+        CONF_PASSKEY: None,
+        CONF_USERNAME: "admin",
+        CONF_PASSWORD: "secret",
+    }
+
+
+async def test_zeroconf_discovery_connection_error(
+    hass: HomeAssistant,
+    mock_bsblan: MagicMock,
+    zeroconf_discovery_info: ZeroconfServiceInfo,
+) -> None:
+    """Test connection error during zeroconf discovery shows the correct form."""
+    mock_bsblan.device.side_effect = BSBLANConnectionError
+
+    discovery_info = zeroconf_discovery_info
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_ZEROCONF},
+        data=discovery_info,
+    )
+
+    assert result.get("type") is FlowResultType.FORM
+    assert result.get("step_id") == "discovery_confirm"
+
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={
+            CONF_PASSKEY: "1234",
+            CONF_USERNAME: "admin",
+            CONF_PASSWORD: "admin1234",
+        },
+    )
+
+    assert result2.get("type") is FlowResultType.FORM
+    assert result2.get("step_id") == "discovery_confirm"
+    assert result2.get("errors") == {"base": "cannot_connect"}
+
+
+async def test_zeroconf_discovery_doesnt_update_host_port_on_existing_entry(
+    hass: HomeAssistant,
+    mock_bsblan: MagicMock,
+    zeroconf_discovery_info: ZeroconfServiceInfo,
+) -> None:
+    """Test that discovered devices don't update host/port of existing entries."""
+    # Create an existing entry with different host/port
+    entry = MockConfigEntry(
+        domain=DOMAIN,
+        data={
+            CONF_HOST: "192.168.1.100",  # Different IP
+            CONF_PORT: 8080,  # Different port
+            CONF_USERNAME: "admin",
+            CONF_PASSWORD: "admin1234",
+        },
+        unique_id="00:80:41:19:69:90",
+    )
+    entry.add_to_hass(hass)
+
+    # Mock zeroconf discovery of the same device but with different IP/port
+    discovery_info = zeroconf_discovery_info
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_ZEROCONF},
+        data=discovery_info,
+    )
+
+    # Should abort because device is already configured
+    assert result["type"] == FlowResultType.ABORT
+    assert result["reason"] == "already_configured"
+
+    # Verify the existing entry was NOT updated with new host/port
```

Comment:
if the zeroconf mac is the same, why don't we update this one? Because apparently the device now advertises on a different IP, so shouldn't we adapt?

Reply:
Yeah, this was changed due to the shared method issue. I’m not sure why, but it should simply update the port and IP with the abort.